ic-task-scheduler/src/scheduler.rs

```diff
@@ -1,86 +1,170 @@
+    tasks_running: TaskQueue,
+    /// Callback to be called to save the current canister state to prevent panicking tasks.
+    on_execution_state_changed_callback: Arc<Box<SaveStateQueryCallback>>,
 }
 
-impl<T: 'static + Task, P: 'static + UnboundedMapStructure<u32, ScheduledTask<T>>> Scheduler<T, P> {
+impl<T, P> Scheduler<T, P>
+where
+    T: 'static + Task,
+    P: 'static + IterableUnboundedMapStructure<u32, ScheduledTask<T>>,
+{
     /// Create a new scheduler.
-    pub fn new(pending_tasks: P) -> Self {
-        Self {
+    ///
+    /// A callback must be passed and it'll be called to save and report the current state of the scheduler.
+    ///
+    /// ATTENTION! In order to prevent scheduler panic this callback should always make a query to a canister endpoint.
+    /// This because the current scheduler state should be saved while running tasks
+    pub fn new(
+        pending_tasks: P,
+        on_execution_state_changed_callback: Box<SaveStateQueryCallback>,
+    ) -> Result<Self> {
+        Ok(Self {
             pending_tasks: Arc::new(Mutex::new(pending_tasks)),
             phantom: std::marker::PhantomData,
-            failed_task_callback: Arc::new(None),
-        }
-    }
-
-    /// Set a callback to be called when a task fails.
-    pub fn set_failed_task_callback<F: 'static + Send + Fn(ScheduledTask<T>, SchedulerError)>(
-        &mut self,
-        cb: F,
-    ) {
-        self.failed_task_callback = Arc::new(Some(Box::new(cb)));
+            on_execution_state_changed_callback: Arc::new(on_execution_state_changed_callback),
+            tasks_to_be_processed: Arc::new(Mutex::new(HeapBTreeMap::new(()))),
+            tasks_running: Arc::new(Mutex::new(HeapBTreeMap::new(()))),
+        })
     }
 
     /// Execute all pending tasks.
     /// Each task is executed asynchronously in a dedicated ic_cdk::spawn call.
     /// This function does not wait for the tasks to complete.
     /// Returns the number of tasks that have been launched.
-    pub fn run(&self) -> Result<u32, SchedulerError> {
-        self.run_with_timestamp(time_secs())
+    pub async fn run(&mut self) -> Result<u32> {
+        self.run_with_timestamp(time_secs()).await
     }
 
-    fn run_with_timestamp(&self, now_timestamp_secs: u64) -> Result<u32, SchedulerError> {
-        let mut to_be_reprocessed = Vec::new();
+    async fn run_with_timestamp(&mut self, now_timestamp_secs: u64) -> Result<u32> {
         let mut task_execution_started = 0;
-        {
-            let mut lock = self.pending_tasks.lock();
-            while let Some(key) = lock.first_key() {
-                let task = lock.remove(&key);
-                drop(lock);
-                if let Some(task) = task {
-                    if task.options.execute_after_timestamp_in_secs > now_timestamp_secs {
-                        to_be_reprocessed.push(task);
+
+        // checks tasks that are still in tasks running (something bad happened in the last cycle)
```

Comment:
(something bad happened in the last cycle) -> (this could indicate a panic happened in the previous run)